lib/semantics/canonicalize-do.cc

```diff
@@ -33,6 +33,30 @@ class CanonicalizationOfDoLoops {
         std::visit(
             common::visitors{
                 [](auto &) {},
+                // Labels on end-stmt of constructs are accepted by f18 for
+                // compatibility purposes, even-though they are technically
+                // not in the same scope as the label-do-stmt.
+                [&](common::Indirection<IfConstruct> &ifConstruct) {
+                  CanonicalizeIfMatch(block, stack, i,
+                      std::get<Statement<EndIfStmt>>(ifConstruct.value().t));
+                },
+                [&](common::Indirection<DoConstruct> &doConstruct) {
+                  CanonicalizeIfMatch(block, stack, i,
+                      std::get<Statement<EndDoStmt>>(doConstruct.value().t));
+                },
+                [&](common::Indirection<CaseConstruct> &caseConstruct) {
+                  CanonicalizeIfMatch(block, stack, i,
+                      std::get<Statement<EndSelectStmt>>(
+                          caseConstruct.value().t));
+                },
+                [&](common::Indirection<SelectRankConstruct> &selectRank) {
+                  CanonicalizeIfMatch(block, stack, i,
+                      std::get<Statement<EndSelectStmt>>(selectRank.value().t));
+                },
+                [&](common::Indirection<SelectTypeConstruct> &selectType) {
+                  CanonicalizeIfMatch(block, stack, i,
+                      std::get<Statement<EndSelectStmt>>(selectType.value().t));
+                },
```

Comment:
Also `associate`, `block`, `forall`, `where`. and `critical`, maybe more.

Reply:
OK, I only took the f95 executable constructs (+ `select type` and `select rank`, because they were "dragged" by `select case` support for simplicity), the rest was rejected by resolve-labels.cc. However, it seems more consistent to accept it on all executable construct end statements (+ forall and where end statements) (pgfortran does so for all the ones it supports).

I have added the suggested plus `change team` and renamed the functions/variables accordingly.

I also added a non-standard feature to control warning on label do loops terminating at statements different than `end do` or `continue`. I can revert this last commit if you think it is not justified.